Rework product tax category storage

ryanreeves-taxjar opened the pull request:
Storing product tax categories as a configuration value won't scale well
as we continue to increase the total number of categories.  Reworking
categories to be stored in their own table eliminates this problem and
will support significantly more categories.

ryanreeves-taxjar committed:
Rework product tax category storage

Storing product tax categories as a configuration value won't scale well
as we continue to increase the total number of categories.  Reworking
categories to be stored in their own table eliminates this problem and
will support significantly more categories.

fastdivision commented on Setup/UpgradeSchema.php:
```diff
@@ -86,5 +86,51 @@ public function upgrade(SchemaSetupInterface $setup, ModuleContextInterface $con
+                    ['identity' => true, 'unsigned' => true, 'nullable' => false, 'primary' => true],
+                    'Product Tax Code'
+                )->addColumn(
+                    'product_tax_code',
```
@ryanreeves-taxjar Should we set up an index for `product_tax_code` since we're loading categories solely by that column in Observer/ImportCategories.php to prevent duplicate categories when importing?

ryanreeves-taxjar replied:
Good idea, added an index to that column.

fastdivision commented on Setup/UpgradeSchema.php:
```diff
@@ -86,5 +86,51 @@ public function upgrade(SchemaSetupInterface $setup, ModuleContextInterface $con
+                    \Magento\Framework\DB\Ddl\Table::TYPE_INTEGER,
+                    null,
+                    ['identity' => true, 'unsigned' => true, 'nullable' => false, 'primary' => true],
+                    'Product Tax Code'
```
Should be `Product Tax Code ID` or `ID`

fastdivision commented on Observer/ImportCategories.php:
```diff
@@ -122,7 +131,16 @@ private function _getCategoryJson()
+        foreach ($categoryJson as $categoryData) {
+            $category = $this->categoryFactory->create();
+
+            $this->categoryResourceModel->load($category, $categoryData['product_tax_code'], 'product_tax_code');
```
@ryanreeves-taxjar Can you add a brief comment above here explaining that we're loading by `product_tax_code` to prevent duplicates?

ryanreeves-taxjar committed:
Add connect/disconnect actions, add table index

Populate new categories from the API during setup.  Delete old
categories when disconnecting from TaxJar.  Add an index to
product_tax_code because it is used for lookups.  Misc code cleanup.

ryanreeves-taxjar committed:
Refactor product tax class admin forms

Refactor the usage of deprecated widget/form classes in favor of ui
components.  Add a better search experience when assigning product tax
categories.

ryanreeves-taxjar committed:
Rename button labels

Change the button labels to match the previous entries.

ryanreeves-taxjar committed:
Add reset button

Add a reset button and clean up the ordering.

ryanreeves-taxjar committed:
Merge remote-tracking branch 'remotes/origin/develop' into INT-45-tax-category-rework

# Conflicts:
#	Setup/UpgradeData.php

ryanreeves-taxjar committed:
Rework product tax category storage

Update setup_version, add trim to names/descriptions pulled from api.

ryanreeves-taxjar committed:
Merge branch 'develop' into INT-45-tax-category-rework

# Conflicts:
#	Setup/UpgradeData.php

ryanreeves-taxjar committed:
Revert refactoring admin forms.  Sort the product tax categories option array.  Bump the setup_version.

ryanreeves-taxjar committed:
Merge branch 'INT-45-tax-category-rework' of https://github.com/taxjar/taxjar-magento2-extension into INT-45-tax-category-rework

 Conflicts:
	Setup/UpgradeData.php

ryanreeves-taxjar commented:
Closing this PR in favor of updated https://github.com/taxjar/taxjar-magento2-extension/pull/98
